unix/w0vncserver/wlr/WlrPixelBuffer.cxx

```diff
@@ -0,0 +1,113 @@
+/* Copyright 2025 Adam Halim for Cendio AB
+ *
+ * This is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this software; if not, write to the Free Software
+ * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307,
+ * USA.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <assert.h>
+#include <stdexcept>
+#include <unistd.h>
+
+#include <wayland-client-core.h>
+#include <wayland-client-protocol.h>
+#include <wayland-client.h>
+
+#include <core/LogWriter.h>
+#include <rfb/VNCServerST.h>
+
+#include "WlrScreencopyManager.h"
+
+#include "../w0vncserver.h"
+#include "../wayland/WOutput.h"
+#include "../wayland/WDisplay.h"
+#include "WlrPixelBuffer.h"
+
+static core::LogWriter vlog("WlrPixelBuffer");
+
+WlrPixelBuffer::WlrPixelBuffer(WDisplay* display, WOutput* output_,
+                               rfb::VNCServer* server_,
+                               std::function<void()> desktopReadyCallback_)
+  : WlrScreencopyManager(display, output_), captureInProgress(false),
+    firstFrame(true), desktopReadyCallback(desktopReadyCallback_),
+    server(server_)
+{
+  captureFrame();
+}
+
+WlrPixelBuffer::~WlrPixelBuffer()
+{
+  if (captureInProgress)
+    server->unblockUpdates();
+}
+
+void WlrPixelBuffer::captureFrame()
+{
+  // We're too slow, skip this frame
+  if (captureInProgress) {
+    // FIXME: We're likely to make two calls to startFrameCapture()
+    // for our first frame. Let's filter out the message the first time
+    // as it is annoying.
+    if (!firstFrame)
+      vlog.debug("Frame capture already in progress, skipping frame");
+    return;
+  }
+
+  if (!firstFrame)
+    server->blockUpdates();
```

Comment:
This is pretty heavy handed. It would be preferred it this is blocked very briefly. Ideally without waiting for something external.